src/coreclr/vm/eetwain.cpp

```diff
@@ -1693,60 +1660,84 @@ PTR_VOID EECodeManager::GetExactGenericsToken(PREGDISPLAY     pContext,
             );
 
     INT32 spOffsetGenericsContext = gcInfoDecoder.GetGenericsInstContextStackSlot();
     if (spOffsetGenericsContext != NO_GENERICS_INST_CONTEXT)
     {
-
-        TADDR taSlot;
-        if (pCodeInfo->IsFunclet())
+        // For funclets we need to recover the establisher frame from
+        // the main function.
+        //
+        // Finally funclets can be called directly from the managed code
+        // in non-exceptional path. In that case the we don't need to do
+        // any adjustment.
+        //
+        // For calls made through CallEHFunclet and CallEHFilterFunclet
+        // we recover the establisher frame address from the frame of the
+        // helper call.
+        //
+        // Presumably profiler callbacks are not generated for funclets
+        // so we can use returnAddress == 0 and skip this.
+        if (returnAddress != 0 &&
+            pCodeInfo->IsFunclet() &&
+            !ExecutionManager::IsManagedCode(returnAddress))
         {
-            INT32 spOffsetPSPSym = gcInfoDecoder.GetPSPSymStackSlot();
-            _ASSERTE(spOffsetPSPSym != NO_PSP_SYM);
-
-#ifdef TARGET_AMD64
-            // On AMD64 the spOffsetPSPSym is relative to the "Initial SP": the stack
-            // pointer at the end of the prolog before and dynamic allocations, so it
-            // can be the same for funclets and the main function.
-            // However, we have a caller SP, so we need to convert
-            baseStackSlot -= pCodeInfo->GetFixedStackSize();
-
-#endif // TARGET_AMD64
-
-            // For funclets we have to do an extra dereference to get the PSPSym first.
-            TADDR newBaseStackSlot = *PTR_TADDR(baseStackSlot + spOffsetPSPSym);
+            SIZE_T offsetOfEstablisherFrameInFuncletSP;
+#if defined(TARGET_AMD64)
+            offsetOfEstablisherFrameInFuncletSP = 0;
+#elif defined(TARGET_X86)
+            // CallEHFunclet frame size + return address + 16 (5th parameter)
+#ifdef UNIX_X86_ABI
+            offsetOfEstablisherFrameInFuncletSP = 48;
+#else
+            offsetOfEstablisherFrameInFuncletSP = 36;
+#endif
+#elif defined(TARGET_ARM64) || defined(TARGET_RISCV64) || defined(TARGET_LOONGARCH64)
+            // Stored past FP/LR pair
+            offsetOfEstablisherFrameInFuncletSP = 16;
+#elif defined(TARGET_ARM)
+            offsetOfEstablisherFrameInFuncletSP = 0;
+#elif defined(TARGET_WASM)
+            _ASSERTE(!"CallFunclet for WASM not implemented yet");
+            offsetOfEstablisherFrameInFuncletSP = 0;
+#else
+            PORTABILITY_ASSERT("offsetOfEstablisherFrameInFuncletSP");
+#endif
 
+            // Recover the establisher frame (InitialSP/CallerSP) from the funclet
+            // caller.
+            baseStackSlot += offsetOfEstablisherFrameInFuncletSP;
+            baseStackSlot = *(SIZE_T*)baseStackSlot;
 #ifdef TARGET_AMD64
             // On AMD64 the PSPSym stores the "Initial SP": the stack pointer at the end of
```

Comment:
It seems this comment is obsolete as it references PSPSym that the PR is removing.

Reply:
Right. This comment is wrong, as are some of the parts of the modified documentation. I'm going to address it once I get the functional parts working properly.